loganalyzer/custom_analyzers/DangerAnalyzer.py

```diff
@@ -83,78 +90,161 @@ def __init__(self, game):
-                        self.danger_teammate_dis[j+1] = temp_dis
-
-                        temp_x = self.danger_teammate_pos[j][1]
-                        temp_y = self.danger_teammate_pos[j][2]
-                        self.danger_teammate[j][1] = self.danger_teammate[j+1][1]
-                        self.danger_teammate[j+1][1] = temp_x
-                        self.danger_teammate[j][2] = self.danger_teammate[j+1][2]
-                        self.danger_teammate[j+1][2] = temp_y
+                    temp_angle = self.temp_opponent_angle[j-1]
+                    self.temp_opponent_angle[j-1] = self.temp_opponent_angle[j]
+                    self.temp_opponent_angle[j] = temp_angle
+
+                    temp_x = self.temp_opponent_pos[j-1][0]
+                    temp_y = self.temp_opponent_pos[j-1][1]
+                    self.temp_opponent_pos[j-1][0] = self.temp_opponent_pos[j][0]
+                    self.temp_opponent_pos[j][0] = temp_x
+                    self.temp_opponent_pos[j-1][1] = self.temp_opponent_pos[j][1]
+                    self.temp_opponent_pos[j][1] = temp_y
         
-                    
-        pass
+        self.danger_opponent_dis.append(self.temp_opponent_dis[0:3])
+        self.danger_opponent_pos.append(self.temp_opponent_pos[0:3])
+        self.danger_opponent_angle.append(self.temp_opponent_angle[0:3])
+
+    def teammate_data(self, cycle, i):
+        for agent in self.game.left_team.agents:
+            dx = agent.data[cycle-i]['x'] - self.game.ball_pos[cycle-i]['x']
+            dy = agent.data[cycle-i]['y'] - self.game.ball_pos[cycle-i]['y']
+            distance = math.sqrt(dx**2 + dy**2)
+            self.temp_teammate_dis.append(distance)
+            self.temp_teammate_pos[len(self.temp_teammate_dis) - 1][0] = agent.data[cycle-i]['x']
+            self.temp_teammate_pos[len(self.temp_teammate_dis) - 1][1] = agent.data[cycle-i]['y']
+
+    def closest_teammate_data(self):
+        for k in range(4):
+            for j in range(len(self.temp_teammate_dis) - 1, k, -1):
+                if self.temp_teammate_dis[j] < self.temp_teammate_dis[j-1]:
+                    temp_dis = self.temp_teammate_dis[j-1]
+                    self.temp_teammate_dis[j-1] = self.temp_teammate_dis[j]
+                    self.temp_teammate_dis[j] = temp_dis
+
+                    temp_x = self.temp_teammate_pos[j-1][0]
+                    temp_y = self.temp_teammate_pos[j-1][1]
+                    self.temp_teammate_pos[j-1][0] = self.temp_teammate_pos[j][0]
+                    self.temp_teammate_pos[j][0] = temp_x
+                    self.temp_teammate_pos[j-1][1] = self.temp_teammate_pos[j][1]
+                    self.temp_teammate_pos[j][1] = temp_y
+        
+        self.danger_teammate_dis.append(self.temp_teammate_dis[0:4])
+        self.danger_teammate_pos.append(self.temp_opponent_pos[0:4])
+
+    def check_danger_param(self, cycle):
+        if (self.check_cycle(cycle)):
```

Comment:
Não sei muito bem aonde ficaria melhor, mas acho que vale a pena deixar registrado em algum lugar (seja por meio de um comentário, ou por meio de documentação da função "check_cycle", ou algum outro lugar) o porquê de conferir isso de ver se os ciclos estão nesse intervalo definido e por que são 5 ciclos, pra quem for mexer no código no futuro ficar mais fácil de entender e não parecer só um "magic number"